compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6486,6 +6488,20 @@ public void visit(BLangOrderByClause orderByClause, AnalyzerData data) {
         }
     }
 
+    @Override
+    public void visit(BLangGroupByClause groupByClause, AnalyzerData data) {
+        groupByClause.env = data.commonAnalyzerData.queryEnvs.peek();
+        for (GroupingKeyNode groupingKey : groupByClause.getGroupingKeyList()) {
+            BLangNode groupingKeyNode = (BLangNode) groupingKey.getGroupingKey();
+            if (groupingKeyNode.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+                checkExpr((BLangSimpleVarRef) groupingKeyNode, groupByClause.env, data);
+            } else {
+                semanticAnalyzer.analyzeNode(groupingKeyNode, groupByClause.env,
+                        data.commonAnalyzerData);
+            }
+        }
+    }
```

Comment:
In order to analyze grouping keys, we are doing two checks here. One is inside `getGroupingKey()` and another one is the node kind check. 

If we follow the below pattern, we can reduce that into a single check.

```java
    @Override
    public void visit(BLangGroupByClause groupByClause, AnalyzerData data) {
        groupByClause.env = data.commonAnalyzerData.queryEnvs.peek();
        for (BLangGroupingKey groupingKey : groupByClause.groupingKeyList) {
            if (groupingKey.variableRef != null) {
                checkExpr(groupingKey.variableRef, groupByClause.env, data);
            } else {
                semanticAnalyzer.analyzeNode(groupingKey.variableDef, groupByClause.env,
                        data.commonAnalyzerData);
            }
        }
    }
```

Reply:
We can do this change if we are changing [[1]](https://github.com/ballerina-platform/ballerina-lang/pull/37555#discussion_r959141680)